sklearn/tests/test_metadata_routing.py

```diff
@@ -239,6 +240,28 @@ class InvalidObject:
         process_routing(InvalidObject(), "fit", groups=my_groups)
 
 
+def test_process_routing_empty_params_get_with_default():
+    empty_params = {}
+    routed_params = process_routing(ConsumingClassifier(), "fit", **empty_params)
+
+    # Behaviour should be an empty dictionary returned for each method when retrieved.
+    for method in METHODS:
+        params_for_method = routed_params[method]
+
+        # An empty dictionary for each method
+        assert isinstance(params_for_method, dict)
+        assert len(params_for_method) == 0
+
+        # This behaviour should be equivalent with using `get` with no default
+        assert routed_params.get(method) == params_for_method
+
+        # However, with a default, should return that instead.
+        assert routed_params.get(method, default="default") == "default"
+
+        # This would fail due to use of `if not default` instead of `if default is None`
+        # assert routed_params.get(method, default=[]) == []
```

Comment:
Please advise on this part. I raised it earlier in the PR but there's been no comment there.

It's hard to find exact usage without types but by searching for `.get(` across the repo and I found a few places there is `params.get("x", {})`. It's hard to tell if `params` is a `dict` or an `EmptyRequest`.

It might not be a problem yet but I feel like this could silently cause hard to debug issues in the future, especically in cases where you expected a `{}` but instead got a `Bunch(**{method: {} for method in METHODS})`.

First recommendation is use a sentinel value to indicate nothing was passed in. Similar to [how `more_itertools` works with defaults](https://github.com/more-itertools/more-itertools/blob/b4e68d645a9815056678fa1795423eb18c6a2d44/more_itertools/more.py#L187-L211). This allows things to work like so:

```python
default_bunch = params.get("x")
none_value = params.get("x", default=None)
list_value = params.get("x", default=[])
``` 

Second recommendation if you do not wish to introduce a sentinel pattern is just to use an explicit `if default is None` check instead of implicit falsyness. However this might not work as expected:

```python
default_bunch = params.get("x")
list_value = params.get("x", default=[])

if params.get("x", default=None) is None:
   # This can never happen
```

Reply:
IMO, the most intuitive approach here is the sentinel value. Basically, not passing anything will always return a `Bunch`. Setting `default` will return the type of `default`.

Such semantic is not surprising and expected. Right now, having `None` returning a Bunch is indeed surprising.

I don't know what @adrinjalali thinks?

Reply:
The behavior should be like a dictionary, when passed the default and the key doesn't exist, we return the default. In this case, I wonder if we should ignore default completely (there only to immitate `dict`), and always return the empty routing list. Afterall, the whole point of this class is to return an empty routed_params object.

Reply:
I'm not opposed at removing the `default` param. However, we would need to change the pattern:

```python
params.get("fit", default={})
```

that is used in the pipeline for instance.

Reply:
I wouldn't remove it, it just always return empty and ignores `default`. The `default` needs to be there to mimic a `dict().get`